File: 0739-daily-temperatures/0739-daily-temperatures.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* dailyTemperatures(int* temperatures, int temperaturesSize, int* returnSize) {
    int* answer = (int*)calloc(temperaturesSize, sizeof(int));
    *returnSize = temperaturesSize;

    for (int i = temperaturesSize - 2; i >= 0; i--) {
        int j = i + 1;

        while (j < temperaturesSize && temperatures[j] <= temperatures[i]) {
            if (answer[j] > 0) {
                j += answer[j];  // Skip forward
            } else {
                j = temperaturesSize;  // No warmer day
            }
        }

        if (j < temperaturesSize) {
            answer[i] = j - i;
        }
    }

    return answer;
}

File: 0739-daily-temperatures/0739-daily-temperatures.java
//Traverse from right, O(n) because of skips (memoization- reusing calculated values)
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];

        for (int i = n - 2; i >= 0; i--) {
            int j = i + 1;

            while (j < n && temperatures[j] <= temperatures[i]) {
                if (answer[j] > 0) {
                    j += answer[j]; // skip ahead to next known warmer day when there is answer on answer[i]
                } else {
                    j = n; // no warmer days ahead
                }
            }

            if (j < n) {
                answer[i] = j - i;
            }
        }

        return answer;
    }
    
}